### DISCUSSION TITLE
Different roots in header use different algorithms, and not Eth2.0-compatible

### DISCUSSION THREAD

**Discussion by chaoticlonghair** [axonweb3/axon#1342] [General]
### Current Behavior

Different roots in block header use different algorithms and different default values.

- The default value for `TrieMerkle.root_hash`, when there are no data, is `hash(rlp(null))`.

  ```rust
            root: Node::Empty,
            root_hash: hasher.digest(rlp::NULL_RLP.as_ref()),
  ```
  
  Ref: [citahub/cita-trie](https://github.com/citahub/cita-trie/blob/6dd02de9f27d2384e083916dfceb30fae322bdd6/src/trie.rs#L217-L218)

- Transactions root for empty list is `hash(rlp(null))`.

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/consensus/src/engine.rs#L71-L79

  I checked that `RLP_NULL == hash(rlp(null))`, and a unit test will be added later.

- Receipts root for empty list is `hash(rlp(null))`.

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/executor/src/lib.rs#L191-L195

- Signed transactions hash for empty list is `hash(rlp([]))`.

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/consensus/src/engine.rs#L86

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/consensus/src/util.rs#L17-L19

- Default value for `metadata_root` and `cell_header_root` is `0x0000..0000`.

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/executor/src/adapter/mod.rs#L287-L293

  https://github.com/axonweb3/axon/blob/4ef2f5608633d8a60bfd50ab337b4965de3b714b/core/executor/src/adapter/mod.rs#L189-L213

Ethereum has defined the following roots, as of today (2023-08-15):

- `transactions_root`

  - Eth1.0: Default value is `hash(rlp(null))`.

  - Eth2.0: Default value is `ssz_root(null)`.

  Ref: [EIP-6404](https://eips.ethereum.org/EIPS/eip-6404).

- `receipts_root`

  Default value is `hash(rlp(null))`.

### Expected Behavior

Default values:
- `transactions_root`: `hash(rlp(null))`
- `receipts_root`: `hash(rlp(null))`.
- `signed_txs_hash`: `hash(rlp([]))` 
  Different with above roots, the inner is an empty list.
- `metadata_root`: `H256::default()`.
  `metadata_root` should be treated as an optional value:
  - `H256::default()` means it's null.
  - `hash(rlp(null))` is the default value when it's not null.
- `cell_header_root`: `H256::default()`.
  Same as `metadata_root`.

### OS

Not Associated

### Kernel

Not Associated

### Anything else?

Since Axon uses MPT root for all roots, but Ethereum uses SSZ root for all roots.

Maybe Axon has to update the `README.md` as following:

```diff
- Axon is compatible with Ethereum so that all of the develop utilities can be used on it directly.
+ Axon is compatible with Ethereum 1.0 so that all of the develop utilities can be used on it directly.
```

**Comment by EthAnalysis** [comment]
Did this kill the Axon project? I noticed no developer announcements or discussion since 2 months after this post in 2023. So basically, I am believing that AXON is only compatible with the old Eth and not the new Eth2.0. I am guessing this means that Axon and all L2's that fork from Axon would have to be upgraded, not only for Eth, but also IBC, whenever those protocols execute a hardfork, and if the Nervos team stops developing, as it seems they have on Axon, then, all L2's/L3's, and all Dapps, and TVL on all are F*#%^ed! (Like what happened to Godwoken).
